[pt2.0/inductor] Fix race in cache dir across ranks on the same host

xw285cornell opened the pull request:
Summary:
It looks we have some race in the cache directory for triton codegen, when we have multiple processes on the same host:
1. Rank A and B cannot find the code in cache (/tmp/uid/triton/cache) and start compilation separately
2. Most of the times the codegen is the same; but rarely it may produce different llir and different shared memory (in our case it's 544 and 2560, both are valid for the llir/ptx generated). See repro D42584580
3. They both write the compiled so and metadata into the local cache folder, with the same directory name (same hash, without considering device id). There will be a race here even if they grab the file lock, because it only locks each file but not the entire transaction
4. We then load the so and meta data back from the file. What happens can be we load so from rank A and shared memory from rank B and they mismatch.

Test Plan:
Run the faulty program to double check
```
[trainer5]: cache dir: /tmp/root/4951/triton/cache/198ef4405d2e525acd20d5c2d01ad099
[trainer1]: cache dir: /tmp/root/4947/triton/cache/198ef4405d2e525acd20d5c2d01ad099
```

Differential Revision: D42619405



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

ngimel commented:
Oh wait, if pid is part of cache dir name, cache won't be reused for a different process launched later? That would slow down repeat runs a lot

xw285cornell commented:
> Oh wait, if pid is part of cache dir name, cache won't be reused for a different process launched later? That would slow down repeat runs a lot

Sounds good, moving to device-based directory

xw285cornell committed:
[pt2.0/inductor] Fix race in cache dir across ranks on the same host (#92664)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/92664

It looks we have some race in the cache directory for triton codegen, when we have multiple processes on the same host:
1. Rank A and B cannot find the code in cache (/tmp/uid/triton/cache) and start compilation separately
2. Most of the times the codegen is the same; but rarely it may produce different llir and different shared memory (in our case it's 544 and 2560, both are valid for the llir/ptx generated). See repro D42584580
3. They both write the compiled so and metadata into the local cache folder, with the same directory name (same hash, without considering device id). There will be a race here even if they grab the file lock, because it only locks each file but not the entire transaction
4. We then load the so and meta data back from the file. What happens can be we load so from rank A and shared memory from rank B and they mismatch.

Test Plan:
Run the faulty program to double check
```

[trainer3]:[2023-01-20 14:08:59,482] torch._inductor.triton_ops.autotune: [INFO] Triton cache directory: /tmp/root/3/triton/cache
[trainer7]:[2023-01-20 14:08:59,691] torch._inductor.triton_ops.autotune: [INFO] Triton cache directory: /tmp/root/7/triton/cache
```

Reviewed By: bertmaher, brad-mengchi, jianyuh, yuguo68, ngimel, eeggl, jiawenliu64

Differential Revision: D42619405

fbshipit-source-id: a0b94f50d48476ff44add111ea6da121a53187d4

jansel reviewed (changes requested):
We also set TRITON_CACHE_DIR here:
https://github.com/pytorch/pytorch/blob/896b6d8768bceff6f9ea6602b3e6bdd517462088/torch/_inductor/codecache.py#L510-L514
and again in another fbcode internal location.

jansel reviewed (approved):
Approving to mitigate S318096, @xw285cornell  will forward fix added duplication.

cchan commented:
This seems to be the wrong fix - after this patch, caches are never reused, and TORCHINDUCTOR_CACHE_DIR / TRITON_CACHE_DIR are ignored.

Proposal: revert this, and fix the problem at the top of this PR by writing to a per-process `/tmp/{pid}/cache_dir`, and atomically moving this to the TRITON_CACHE_DIR location during cleanup.

xw285cornell commented:
hmm @cchan can you elaborate? I should not have touched TORCHINDUCTOR_CACHE_DIR. TRITON_CACHE_DIR I did ignore whatever was set in the torch/_inductor/codecache.py which I'm cleaning up in https://www.internalfb.com/diff/D42694374. But why would this make it never reuse the cache? I added the device_id into the cache path and that's it. Using pid is not ideal because that'll actually make it not reusing the cache (because every new process will be on a new pid, defeating the purpose of a disk cache).

jansel commented:
I think this PR is causing issues for people who set a custom TORCHINDUCTOR_CACHE_DIR / TRITON_CACHE_DIR.  

Previously if I did:
```
TORCHINDUCTOR_CACHE_DIR=/home/jansel/cache
```

Then `/tmp` would not be used at all.  Now we hardcode `/tmp` usage.  Some systems have odd `/tmp` setups.

cchan commented:
For my usecase, I use this in a cluster, and I'd like to share my `TORCHINDUCTOR_CACHE_DIR` between machines on NFS. I'd prefer not to use /tmp.

jansel commented:
This should be fixed by https://github.com/pytorch/pytorch/pull/92879

xw285cornell commented:
> I think this PR is causing issues for people who set a custom TORCHINDUCTOR_CACHE_DIR / TRITON_CACHE_DIR.
> 
> Previously if I did:
> 
> ```
> TORCHINDUCTOR_CACHE_DIR=/home/jansel/cache
> ```
> 
> Then `/tmp` would not be used at all. Now we hardcode `/tmp` usage. Some systems have odd `/tmp` setups.

ah great point, I forgot about that. Glad that this was fixed in #92879 !
